Add Account dets, Loans, Reservations, Settings, & Staff Management

pashiav opened the pull request:
Added logging in with card number, displaying user account information, loans, reservations, settings, and staff management.

---

**Testing instructions:**
1. Open a new terminal and run: `ssh -L 3307:mysql.eecs.ku.edu:3306 [your_username]@cycle3.eecs.ku.edu`
		`-L`: Local port forwarding
		`3307`: port you want to access the database from locally (localhost:3307)
		`mysql.eecs.ku.edu:3306`: mysql database server and port you want to forward to
		`[your_username]@cycle3.eecs.ku.edu`: ssh in to the cycle server to set up tunnel

Common issue:
**- Error connecting to port 22**
	**Fix**: Switch wifi networks (e.g. from your home network to your hotspot) and disconnect/reconnect to KU Anywhere VPN. Try step 1 again.
	If still not working, stay connected to second network and try again later (~10min).

2. Create a new file named `.env.local` in the `library` folder. Copy the contents of  `.env.example` into `.env.local` and fill out all necessary fields -> Save the file.
```
DB_HOST=127.0.0.1
DB_PORT=3307
DB_USER=mysql_username
DB_PASSWORD=mysql_password
DB_NAME=mysql_database_name
```

3. In a separate terminal, cd to the app root directory (`library`)
4. Run `npm run dev` -> go to http://localhost:3000 to see the web app

---

- [ ] Log in with card numbers & check that invalid card numbers cannot log in
	- [ ] Stay logged in on every page until clicking **Sign Out**
	- [ ] Can sign out while on any page
- [ ] Account Details are accurate
- [ ] Loan information is accurate
	- [ ] Clicking Return updates the Returned Date
	- [ ] Loan is updated in database

idk how @Iwantexpresso implemented the count for Inventory, so I think that this should be implemented on his branch:
- Clicking **Return** causes the item count to go up

- [ ] Reservation information is accurate
	- [ ] **Cancel** reservation button only shows IFF reservation end date is in the future
		- [ ] Reservation is removed in database
- [ ] Account Settings information is accurate
	- [ ] Can update account details
	- [ ] Can delete account
		- [ ] Is prompted with "Are you sure you want to delete your account?" before deletion

Log in with a staff account
- [ ] Search for an account
- [ ] Upload an item to the database

pashiav committed:
add env.example file

pashiav committed:
add validation to to account log ins

pashiav committed:
add accounts page. account info, loans & reservations

combs-a reviewed (changes requested):
- [X] Log in with card numbers & check that invalid card numbers cannot log in
- [X] General testing of Account info
- [X] General testing of Loans
- [X] General testing of Reservations
<hr>
I assume to test loans and reservations right now, we'll add a loan/reservation via the backend?

I have a few thoughts!

1: I do think that we should make it so that you don't log out if you click on the home/account tabs, especially since loans/reservations are going to be through the inventory tab.

Could we cache the person who's logged in and change the home page to a simple welcome message if they are? i.e. instead of having the library card login, just say like "Welcome, [FIRST NAME] [LAST NAME]"?

2: I think ID isn't necessary information to display on the account page since it's just for DB purposes!

For testing the loans/reservations via adding them in the backend, here's the SQL queries to make it easier:

```
INSERT INTO loan(item_id, account_id, loan_out_date, due_date) VALUES([item id here], [account id here], DATE YYYY/MM/DD, DATE YYYY/MM/DD);

UPDATE item SET quantity_available = (quantity_available - 1) WHERE item_id = [ITEM ID HERE];

INSERT INTO reservations(item_id, account_id, reservation_date, reservation_end_date) VALUES([item id here], [account id here], DATE YYYY/MM/DD, DATE YYYY/MM/DD);

UPDATE item SET reservation_amount = (reservation_amount - 1) WHERE item_id = [ITEM ID HERE];
```

Unrelated to this PR, I just realized that reservations doesn't match with the rest of the table naming conventions, so we should probably update that to just say reservation. Before this gets merged we should update that and then update things in here that reference it.

pashiav committed:
add account info to local storage

pashiav commented:
@combs-a Thank you for reviewing!

> I assume to test loans and reservations right now, we'll add a loan/reservation via the backend?

-> Yes! It was supposed to be tested with the Inventory page, but that has not yet been implemented. Please continue to test by manually inserting through phpMyAdmin.


---
**✅ Fixes Ready for Re-review**
- **1:** Card number input only displays on the home page if the user is not logged in. First name and last name is displayed if logged in. Added a sign out button in the navigation bar.

`Before log in` | `After`

<img width="350" alt="Screenshot 2025-04-27 at 4 45 08 PM" src="https://github.com/user-attachments/assets/46c3b2b8-f12d-4f71-a01b-591822d736da" />
<img width="350" alt="Screenshot 2025-04-27 at 4 45 39 PM" src="https://github.com/user-attachments/assets/ad3a3f22-df5a-4bcb-9f61-0b4bd230d125" />

<br>
<br>

- **2:** Removed ID from Accounts page.

---

I updated the testing instructions!

combs-a reviewed (approved):
- [X] Log in with card numbers & check that invalid card numbers cannot log in
	- [X] Stay logged in on every page until clicking **Sign Out**
	- [X] Can sign out while on any page
- [X] Account Details are accurate
- [X] Loan information is accurate
- [X] Reservation information is accurate
<hr>
Didn't run into any issues, this looks great! Thank you for the changes Pashia!

Iwantexpresso commented:
I have been working with this branch and so far everything has been responsive to the update I've introduced in my own personal branch without issues! Live changes from the inventory page to the database are being reflected and displayed correctly for both loans and reservations!

Iwantexpresso reviewed (approved):
just in case let me add my message here since I beleiove my fist attempt was added as a comment:
I have been working with this branch and so far everything has been responsive to the update I've introduced in my own personal branch without issues! Live changes from the inventory page to the database are being reflected and displayed correctly for both loans and reservations!

pashiav committed:
add account type

pashiav committed:
change color of sign out for distinction

pashiav committed:
add staff panel

pashiav committed:
fix landing page view height

pashiav committed:
fix account view height

pashiav committed:
add account type & buttons to cancel & return items

pashiav commented:
@alesan99 @combs-a @Iwantexpresso @hwolong 
Added Account Settings & Staff Panel/Management to the accounts page. I updated the testing instructions!

Note this comment:

> idk how @Iwantexpresso implemented the count for Inventory, so I think that this should be implemented on his branch:
> - Clicking Return causes the item count to go up

combs-a reviewed (changes requested):
- [X] Log in with card numbers & check that invalid card numbers cannot log in
	- [X] Stay logged in on every page until clicking **Sign Out**
	- [X] Can sign out while on any page
- [X] Account Details are accurate
- [X] Loan information is accurate
	- [X] Clicking Return updates the Returned Date
	- [X] Loan is updated in database

idk how @Iwantexpresso implemented the count for Inventory, so I think that this should be implemented on his branch:
- Clicking **Return** causes the item count to go up

- [X] Reservation information is accurate
	- [X] **Cancel** reservation button only shows IFF reservation end date is in the future
		- [x] Reservation is removed in database
- [X] Account Settings information is accurate
	- [X] Can update account details
	- [X] Can delete account
		- [X] Is prompted with "Are you sure you want to delete your account?" before deletion

Log in with a staff account
- [X] Search for an account
- [X] Upload an item to the database

<hr>

This looks pretty good so far!

Some changes that would be good are:
- When an update is made in the account settings menu, can there be an indication? I updated things a few times and I couldn't tell that anything had been updated.
- Also, could the name fields be separated out (so 3 fields for first, middle initial, and last)? I think as-is, someone's pretty likely to put in something like "First M. Last" to try and make sure their initial is in there, but that changes their name to "First M M."
- In updates, instead of entering foreign keys/IDs, would it be possible to instead fetch and build a dropdown list? Or would that be too much trouble?
- Could we potentially put in adding contributors/contributions via this?

The last two are more complicated, so if someone needs to jump in and take it on then I can see what I can do.

This looks really good though, great work Pashia!

Iwantexpresso reviewed (changes requested):
This looks good so far!, howver im encountering a unique issue for the user peter b jones in which his reservation on The Great Gatsby item cannot be cancelled. other than that this PR should be good to go

combs-a commented:
@Iwantexpresso that's intended behavior--the reservation time period ended so you cannot cancel it anymore!

alesan99 commented:
- [x] Log in with card numbers & check that invalid card numbers cannot log in
	- [x] Stay logged in on every page until clicking **Sign Out**
	- [x] Can sign out while on any page
- [x] Account Details are accurate
- [x] Loan information is accurate
	- [ ] ~~Clicking Return updates the Returned Date~~
	- [ ] ~~Loan is updated in database~~
- [x] Reservation information is accurate
	- [x] **Cancel** reservation button only shows IFF reservation end date is in the future
		- [ ] ~~Reservation is removed in database~~
- [x] Account Settings information is accurate
	- [x] Can update account details
	- [ ] ~~Can delete account~~
		- [x] Is prompted with "Are you sure you want to delete your account?" before deletion
Log in with a staff account
- [x] Search for an account
- [ ] Upload an item to the database

Tested most stuff and it seems good

vvvvv

The page scrolling and footer gets messed up when you search up an account.

![image](https://github.com/user-attachments/assets/f9541b89-0802-4fbe-a9f4-c600db547f4b)

Also can't upload this

pashiav committed:
add background color, fix formatting & ui

pashiav commented:
@alesan99 Added a default background color - should be fixed!

combs-a reviewed (approved):
Forgot to come back in and approve this; since we discussed other improvements will be done in other PRs this should be ready to merge whenever!
